src/plugins/uploader.js

```diff
@@ -0,0 +1,156 @@
+"use strict";
+
+const SocketIOFileUploadServer = require("socketio-file-upload/server");
+const Helper = require("../helper");
+const path = require("path");
+const fsextra = require("fs-extra");
+const fs = require("fs");
+const fileType = require("file-type");
+const readChunk = require("read-chunk");
+const crypto = require("crypto");
+const isUtf8 = require("is-utf8");
+
+class Uploader {
+	constructor(client, socket) {
+		const siofu = new SocketIOFileUploadServer();
+
+		siofu.uploadValidator = function(event, callback) {
+			const folder = path.join(Helper.getFileUploadPath(), ".tmp");
+			fsextra.ensureDir(folder, (err) => {
+				if (err) {
+					callback(err);
+				} else {
+					siofu.dir = folder;
+					callback(true);
+				}
+			});
+		};
+
+		siofu.on("complete", (data) => {
+			const randomFileName = handleSaving(data);
+
+			const slug = data.file.name;
+			const url = `uploads/${randomFileName}/${slug}`;
+			client.emit("upload:success", url);
+		});
+
+		siofu.on("error", (data) => {
+			log.error(`Error uploading ${data.error.name}`);
+			log.error(data.error);
+		});
+
+		// maxFileSize is in bytes, but config option is passed in as KB
+		siofu.maxFileSize = Helper.config.uploadMaxFileSize * 1024;
+		siofu.listen(socket);
+
+		// Returns new file name
+		function handleSaving(data) {
+			const tempPath = path.join(Helper.getFileUploadPath(), ".tmp", data.file.name);
+			let randomName, destPath;
+
+			// If file conflicts
+			do {
+				randomName = crypto.randomBytes(8).toString("hex");
+				destPath = path.join(Helper.getFileUploadPath(), randomName.substring(0, 2), randomName);
+			} while (fs.exists(destPath));
+
+			fsextra.moveSync(tempPath, destPath, (err) => {
```

Comment:
Let's use the async version. It should be really simple as adding `callback` and calling it with `randomName` in this move callback.